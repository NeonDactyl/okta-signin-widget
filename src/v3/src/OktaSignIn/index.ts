/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { OktaAuth, OktaAuthOptions } from '@okta/okta-auth-js';
import pick from 'lodash/pick';
import { h, render } from 'preact';

import { version } from '../../package.json';
import { Widget } from '../components/Widget';
import { WidgetProps } from '../types/widget';

export type RenderOptions = {
  el: string;
  clientId?: string;
  redirectUri?: string;
  redirect?: 'always' | 'never';
  authParams?: OktaAuthOptions;
};

export type OktaWidgetEventHandler = {
  (...args: unknown[]): void;
};

export type RenderCallback = {
  (...args: unknown[]): void;
};

export default class OktaSignIn {
  /**
   * Version string
   */
  readonly version: string;

  /**
   * Okta Signin Widget options
   */
  readonly options: WidgetProps;

  /**
   * Instance of OktaAuth client
   */
  readonly authClient: OktaAuth;

  /**
   * Registered event handlers
   */
  readonly events: {
    [eventName: string]: OktaWidgetEventHandler
  };

  constructor(options: WidgetProps) {
    this.version = version;
    this.options = options;

    this.events = {};

    // if authClient is set, authParams are disregarded
    if (options.authClient) {
      // safety check
      if (!(options.authClient instanceof OktaAuth)) {
        throw new Error('expected "authClient" to be an instance of OktaAuth');
      }

      // instance of OktaAuth client
      this.authClient = options.authClient;
    } else {
      const authParams = {
        ...pick(options, [
          'issuer',
          'clientId',
          'redirectUri',
          'state',
          'scopes',
          'flow',
          'codeChallenge',
          'codeChallengeMethod',
          'recoveryToken',
        ]),
        ...options.authParams,
      };

      // set default issuer using baseUrl
      if (!authParams.issuer) {
        authParams.issuer = `${options.baseUrl}/oauth2/default`;
      }

      // instance of OktaAuth client
      this.authClient = new OktaAuth(authParams);

      // add widget version to extended user agent header
      // eslint-disable-next-line no-underscore-dangle
      const userAgent = this.authClient._oktaUserAgent;
      if (userAgent) {
        userAgent.addEnvironment(`okta-signin-widget-${version}`);
      }

      if (options.el) {
        this.renderEl({
          el: options.el,
          clientId: authParams.clientId,
          redirectUri: authParams.redirectUri,
        });
      }
    }
  }

  renderEl(
    options: RenderOptions,
    onSuccess?: RenderCallback,
    onError?: RenderCallback,
  ): Promise<{ tokens: Record<string, string> }> {
    const { el } = options;

    return new Promise<{ tokens: Record<string, string> }>((resolve, reject) => {
      try {
        const target = typeof el === 'string'
          ? document.querySelector(el)
          : el;

        if (target) {
          render(h(Widget, { ...this.options, authClient: this.authClient }), target);
          if (typeof onSuccess === 'function') {
            onSuccess({});
          }
          resolve({
            tokens: {
              idToken: 'idToken',
              accessToken: 'accessToken',
              refreshToken: 'refreshToken',
            },
          });
        } else {
          throw new Error(`could not find element ${el}`);
        }
      } catch (error) {
        if (typeof onError === 'function') {
          onError(error);
        }
        reject(error);
      }
    });
  }

  showSignInToGetTokens(options = {}): Promise<Record<string, unknown>> {
    if (this.authClient.isAuthorizationCodeFlow() && this.authClient.isPKCE()) {
      throw new Error('"showSignInToGetTokens()" should not be used for authorization_code flow. Use "showSignInAndRedirect()" instead');
    }

    return this.renderEl({
      ...this.buildRenderOptions(options),
      redirect: 'never',
    })
      .then((res) => res.tokens);
  }

  showSignInAndRedirect(options = {}): Promise<unknown> {
    return this.renderEl({
      ...this.buildRenderOptions(options),
      redirect: 'always',
    });
  }

  showSignIn(options = {}): Promise<unknown> {
    return this.renderEl(
      this.buildRenderOptions(options),
    );
  }

  // eslint-disable-next-line class-methods-use-this
  before(): void {}

  // eslint-disable-next-line class-methods-use-this
  after(): void {}

  // eslint-disable-next-line class-methods-use-this
  hide(): void {}

  // eslint-disable-next-line class-methods-use-this
  show(): void {}

  // eslint-disable-next-line class-methods-use-this
  remove(): void {}

  // eslint-disable-next-line class-methods-use-this
  getUser(): void {}

  on(eventName: string, eventHandler: OktaWidgetEventHandler): void {
    this.events[eventName] = eventHandler;
  }

  private buildRenderOptions(
    options: WidgetProps & Record<string, string> = {},
  ): RenderOptions {
    const widgetOptions = this.options;
    const authParams: OktaAuthOptions = {
      ...widgetOptions.authParams,
      ...pick(options, [
        'responseType',
        'scopes',
        'state',
        'nonce',
        'idp',
        'idpScope',
        'display',
        'prompt',
        'maxAge',
        'loginHint',
      ]),
    };

    const { el, clientId, redirectUri } = { ...widgetOptions, ...options };
    if (!el) {
      throw new Error('"el" is required');
    }
    if (!clientId) {
      throw new Error('"clientId" is required');
    }
    if (!redirectUri) {
      throw new Error('"redirectUri" is required');
    }
    return {
      el,
      clientId,
      redirectUri,
      authParams,
    };
  }
}
