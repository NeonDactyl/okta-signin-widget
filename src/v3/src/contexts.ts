/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessage, IdxTransaction, OktaAuth } from '@okta/okta-auth-js';
import { createContext } from 'preact';
import { StateUpdater, useContext } from 'preact/hooks';

import {
  FormBag,
  JsonObject,
  Nullish,
  Undefinable,
} from './types';

type IWidgetContext = {
  data: JsonObject;
  authClient: OktaAuth;
  setIdxTransaction: StateUpdater<Nullish<IdxTransaction>>;
  setMessages: StateUpdater<IdxMessage[]>;
  // // TODO: OKTA-502849 - Update param type
  // (RenderSuccessCallback / RenderErrorCallback) once merged into okta-signin-widget
  onSuccessCallback?: (data: JsonObject) => void;
  onErrorCallback?: (data: JsonObject) => void;
  formBag?: FormBag;
  transaction: Nullish<IdxTransaction>;
};

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<Undefinable<T>>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };
