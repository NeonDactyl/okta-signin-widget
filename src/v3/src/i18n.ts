/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { initOktaReactI18n } from './lib/okta-i18n';
import enTranslations from './properties/json/login.json';

initOktaReactI18n({
  debug: false,
  namespace: 'okta-signin-widget',
  supportedLanguages: ['en', 'cs', 'da', 'de', 'el', 'es', 'fi', 'fr', 'hu', 'id', 'it', 'ja', 'ko', 'ms', 'nb',
    'nl_NL', 'ok_PL', 'pl', 'pt_BR', 'ro', 'ru', 'sv', 'th', 'tr', 'uk', 'vi', 'zh_CN', 'zh_TW'],
  fallbackLanguage: 'en',
  translations: {
    en: enTranslations,
  },
  backendOptions: {
    baseUri: '',
    rewrite: (file) => file,
  },
  languageDetectorOptions: {
    localeCookieName: 'okta_user_lang',
  },
});
