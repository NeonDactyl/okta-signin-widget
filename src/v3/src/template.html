<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title><% preact.title %></title>
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <link rel="apple-touch-icon" href="/static/icons/apple-touch-icon.png">
    <% preact.headEnd %>
  </head>
  <body>
    <div id="signin-container"></div>
    <% preact.bodyEnd %>
  </body>
</html>
