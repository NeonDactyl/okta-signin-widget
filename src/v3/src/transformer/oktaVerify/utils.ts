/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { UISchemaElement } from '@jsonforms/core';
import { IdxAuthenticator } from '@okta/okta-auth-js';
import set from 'lodash/set';
import {
  DescriptionElement,
  ListElement,
  QRCodeElement,
  StepperNavButtonConfigAttrs,
  StepperNavButtonConfigDirection,
} from 'src/types';

export const appendDescriptionElements = (
  elements: UISchemaElement[],
  authenticator?: IdxAuthenticator,
): void => {
  if (!authenticator?.contextualData) {
    return;
  }

  const { contextualData } = authenticator;

  if (contextualData.qrcode) {
    const { href } = contextualData.qrcode;
    const { displayName } = authenticator;
    elements.push({
      type: 'List',
      options: {
        items: [
          'oie.enroll.okta_verify.qrcode.step1',
          'oie.enroll.okta_verify.qrcode.step2',
          'oie.enroll.okta_verify.qrcode.step3',
        ],
        type: 'ordered',
      },
    } as ListElement);

    elements.push({
      type: 'QRCode',
      options: {
        label: displayName,
        data: href,
      },
    } as QRCodeElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { email } = contextualData;
  if (email) {
    elements.push({
      type: 'Description',
      options: {
        content: 'next.enroll.okta_verify.email.info',
        contentParams: [email],
      },
    } as DescriptionElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { phoneNumber } = contextualData;
  if (phoneNumber) {
    elements.push({
      type: 'Description',
      options: {
        content: 'next.enroll.okta_verify.sms.info',
        contentParams: [phoneNumber],
      },
    } as DescriptionElement);
  }
};

export const createNavButtonConfig = (hasQrCode: boolean): {
  navButtonsConfig: Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
} => {
  const stepperNavBtnConfig = {
    navButtonsConfig: {
      next: {
        variant: 'clear',
        label: hasQrCode
          ? 'enroll.totp.cannotScan'
          : 'next.enroll.okta_verify.switch.channel.link.text',
      },
    } as Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
  };

  if (hasQrCode) {
    set(stepperNavBtnConfig, 'navButtonsConfig.prev', {
      variant: 'clear',
      label: 'next.enroll.okta_verify.switch.channel.link.text',
    } as StepperNavButtonConfigAttrs);
  }

  return stepperNavBtnConfig;
};

export const getTitleKey = (selectedChannel?: string): string => {
  switch (selectedChannel) {
    case 'email':
      return 'oie.enroll.okta_verify.setup.email.title';
    case 'sms':
      return 'oie.enroll.okta_verify.setup.sms.title';
    default:
      return 'oie.enroll.okta_verify.setup.title';
  }
};
