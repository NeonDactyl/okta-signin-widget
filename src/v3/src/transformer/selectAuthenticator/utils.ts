/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import {
  IdxOption,
  IdxRemediationValue,
} from '@okta/okta-auth-js/lib/idx/types/idx-js';

import { AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP, AUTHENTICATOR_KEY } from '../../constants';
import {
  AuthenticatorOptionValue,
  Nullish,
  Option,
  Undefinable,
} from '../../types';

const getAuthenticatorOption = (
  options: IdxOption[],
  authenticatorKey: string,
): Undefinable<IdxOption> => options?.find(
  ({ relatesTo }) => relatesTo?.key === authenticatorKey,
);

const getOptionValue = (
  inputs: Input[],
  key: string,
): Nullish<Input> => inputs?.find(
  ({ name }) => name === key,
);

const buildOktaVerifyOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): Option<AuthenticatorOptionValue>[] => {
  const ovRemediation = getAuthenticatorOption(options, AUTHENTICATOR_KEY.OV);
  const id = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'id')?.value;
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType || !methodType.options?.length) {
    return [];
  }

  return methodType.options.map((option) => ({
    key: AUTHENTICATOR_KEY.OV,
    label: option.label,
    description: isEnroll && AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.OV],
    value: {
      label: isEnroll
        ? 'oie.enroll.authenticator.button.text'
        : 'oie.verify.authenticator.button.text',
      methodType: option.value,
      id,
    },
  } as Option<AuthenticatorOptionValue>));
};

const getAuthenticatorDescription = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): Undefinable<string> => {
  if (!authenticatorKey) {
    return undefined;
  }

  if (isEnroll) {
    return AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[authenticatorKey];
  }

  if (authenticatorKey === AUTHENTICATOR_KEY.PHONE) {
    return getAuthenticatorOption(
      options,
      AUTHENTICATOR_KEY.PHONE,
    // @ts-ignore OKTA-499921 (profile missing from relatesTo interface)
    )?.relatesTo?.profile?.phoneNumber;
  }
  return undefined;
};

const getOnPremDescriptionParams = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): Undefinable<string[]> => {
  if (!isEnroll || authenticatorKey !== AUTHENTICATOR_KEY.ON_PREM) {
    return undefined;
  }

  const vendorName = getAuthenticatorOption(
    options,
    AUTHENTICATOR_KEY.ON_PREM,
  )?.relatesTo?.displayName || 'oie.on_prem.authenticator.default.vendorName';
  return [vendorName];
};

const formatAuthenticatorOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): Option<AuthenticatorOptionValue>[] => (
  options.map((option: IdxOption) => {
    const authenticatorKey = option.relatesTo?.key as string;
    const id = getOptionValue(option?.value as Input[], 'id')?.value;
    const methodType = getOptionValue(option?.value as Input[], 'methodType')?.value;
    const enrollmentId = getOptionValue(option?.value as Input[], 'enrollmentId')?.value;

    return {
      key: authenticatorKey,
      label: option.label,
      description: getAuthenticatorDescription(
        options,
        authenticatorKey,
        isEnroll,
      ),
      descriptionParams: getOnPremDescriptionParams(
        options,
        authenticatorKey,
        isEnroll,
      ),
      value: {
        key: authenticatorKey,
        label: isEnroll
          ? 'oie.enroll.authenticator.button.text'
          : 'oie.verify.authenticator.button.text',
        id,
        methodType,
        enrollmentId,
      } as AuthenticatorOptionValue,
    };
  })
);

const getAuthenticatorOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): Option<AuthenticatorOptionValue>[] => {
  const formattedOptions = formatAuthenticatorOptions(options, isEnroll);

  // appending OV options back to its original spot
  const ovOptions = buildOktaVerifyOptions(options, isEnroll);
  if (ovOptions.length && options?.length) {
    const ovIndex = options.findIndex(({ relatesTo }) => relatesTo?.key === AUTHENTICATOR_KEY.OV);
    formattedOptions.splice(ovIndex, 1, ...ovOptions);
  }

  return formattedOptions;
};

export const getOVMethodTypeAuthenticatorOptions = (
  options?: IdxOption[],
): Option<AuthenticatorOptionValue>[] => {
  if (!options) {
    return [];
  }

  return options.map((option) => ({
    key: AUTHENTICATOR_KEY.OV,
    label: option.label,
    value: {
      label: 'oie.verify.authenticator.button.text',
      methodType: option.value as string,
    },
  }));
};

export const isOnlyPushWithAutoChallenge = (
  authenticatorRemediation?: IdxRemediationValue,
): boolean => {
  const methodType = (authenticatorRemediation?.form?.value)
    ?.find(({ name }) => name === 'methodType');
  const autoChallenge = (authenticatorRemediation?.form?.value)
    ?.find(({ name }) => name === 'autoChallenge');

  return typeof autoChallenge !== 'undefined'
    && methodType?.options?.length === 1
    && methodType.options[0].value === 'push';
};

export const getAuthenticatorVerifyOptions = (
  authenticatorOptions: IdxOption[],
): Option<AuthenticatorOptionValue>[] => getAuthenticatorOptions(
  authenticatorOptions,
);

export const getAuthenticatorEnrollOptions = (
  authenticatorOptions: IdxOption[],
): Option<AuthenticatorOptionValue>[] => getAuthenticatorOptions(
  authenticatorOptions,
  true,
);
