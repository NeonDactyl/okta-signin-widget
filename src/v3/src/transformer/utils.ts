/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ControlElement } from '@jsonforms/core';
import { IdxTransaction } from '@okta/okta-auth-js';
import { IdxRemediation } from '@okta/okta-auth-js/lib/idx/types/idx-js';

import { FormBag, LayoutType, RequiredKeys } from '../types';

export const createForm = (): FormBag => ({
  schema: {
    type: 'object',
    properties: {},
    required: [],
  },
  uischema: {
    type: LayoutType.VERTICAL,
    elements: [],
  },
  data: {},
});

export const removeUIElementWithScope = (
  scope: string,
  elements: ControlElement[],
): ControlElement[] => (
  elements.filter((element) => (
    scope !== element.scope
  ))
);

export const getUIElementWithScope = (
  scope: string,
  elements: ControlElement[],
): ControlElement | undefined => (
  elements.find((element) => (
    scope === element.scope
  ))
);

export const isNextStepAvailable = (
  transaction: IdxTransaction,
): transaction is RequiredKeys<IdxTransaction, 'nextStep'> => !!transaction.nextStep;

export const remediationContainsStep = (
  neededToProceed: IdxRemediation[],
  step: string,
): boolean => neededToProceed.some((remediation) => remediation.name === step);

export const getCurrentTimestamp = (): number => (new Date().getTime());
