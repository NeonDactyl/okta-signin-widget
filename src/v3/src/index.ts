/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import './i18n';

import OktaSignIn from './OktaSignIn';

if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
  import('./mocks/browser')
    .then(({ getWorker }) => getWorker())
    .then((worker) => worker?.start());
}

if (typeof window !== 'undefined') {
  // @ts-ignore OKTA-487668
  window.OktaSignIn = OktaSignIn;
}
