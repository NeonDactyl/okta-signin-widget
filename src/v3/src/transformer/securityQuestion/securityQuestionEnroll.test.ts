/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ControlElement, RuleEffect } from '@jsonforms/core';
import { getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import { FormBag } from 'src/types';

import { transformSecurityQuestionEnroll } from '.';

describe('SecurityQuestionEnroll Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  let formBag: FormBag;
  beforeEach(() => {
    formBag = {
      envelope: {},
      data: {},
      schema: {},
      uischema: {
        type: 'VerticalLayout',
        elements: [{
          type: 'Control',
          scope: '#/properties/answer',
        } as ControlElement],
      },
    };
  });

  it('should create security question enrollment UI elements', () => {
    transaction.nextStep = {
      name: '',
      authenticator: {
        contextualData: {
          questions: [{
            question: 'What is love?',
            questionKey: 'eternal',
          }],
        },
        id: '',
        displayName: '',
        key: '',
        type: '',
        methods: [],
      },
    };
    const updatedFormBag = transformSecurityQuestionEnroll(transaction, formBag);

    expect(updatedFormBag).toMatchSnapshot();

    expect(updatedFormBag.uischema.elements.length).toBe(5);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');

    // choice element
    expect(updatedFormBag.uischema.elements[1].type).toBe('Control');
    expect(updatedFormBag.uischema.elements[1].options?.format).toBe('radio');
    expect(updatedFormBag.uischema.elements[1].options?.choices[0].key).toBe('predefined');
    expect(updatedFormBag.uischema.elements[1].options?.choices[1].key).toBe('custom');

    // customQuestion element
    expect(updatedFormBag.uischema.elements[2].type).toBe('Control');
    expect(updatedFormBag.uischema.elements[2].rule).toEqual(expect.objectContaining({
      effect: RuleEffect.SHOW,
      condition: {
        scope: expect.any(String),
        schema: {
          const: 'custom',
        },
      },
    }));

    // predefinedQuestions element
    expect(updatedFormBag.uischema.elements[3].type).toBe('Control');
    expect(updatedFormBag.uischema.elements[3].options?.format).toBe('dropdown');
    expect(updatedFormBag.uischema.elements[3].options?.choices).toStrictEqual([{
      key: 'eternal',
      value: 'What is love?',
    }]);
    expect(updatedFormBag.uischema.elements[3].rule).toEqual(expect.objectContaining({
      effect: RuleEffect.SHOW,
      condition: {
        scope: expect.any(String),
        schema: {
          const: 'predefined',
        },
      },
    }));

    // answer element
    expect(updatedFormBag.uischema.elements[4].type).toBe('Control');
    expect(updatedFormBag.uischema.elements[4].options?.format).toBe('password');

    // ensure default questionType selection
    expect(updatedFormBag.data.questionType).toBe('predefined');

    // ensure schema enum for questionType and questionKey are correct
    expect(updatedFormBag.schema.properties?.questionType.enum).toEqual(['predefined', 'custom']);
    expect(updatedFormBag.schema.properties?.questionKey.enum).toEqual(['custom', 'eternal']);
  });
});
