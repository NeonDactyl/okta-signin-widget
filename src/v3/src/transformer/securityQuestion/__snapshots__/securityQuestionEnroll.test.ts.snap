// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecurityQuestionEnroll Tests should create security question enrollment UI elements 1`] = `
Object {
  "data": Object {
    "questionType": "predefined",
  },
  "envelope": Object {},
  "schema": Object {
    "properties": Object {
      "questionKey": Object {
        "enum": Array [
          "custom",
          "eternal",
        ],
        "type": "string",
      },
      "questionType": Object {
        "enum": Array [
          "predefined",
          "custom",
        ],
        "type": "string",
      },
    },
  },
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.security.question.enroll.title",
        },
        "type": "Title",
      },
      Object {
        "label": false,
        "options": Object {
          "choices": Array [
            Object {
              "key": "predefined",
              "value": "oie.security.question.questionKey.label",
            },
            Object {
              "key": "custom",
              "value": "oie.security.question.createQuestion.label",
            },
          ],
          "format": "radio",
        },
        "scope": "#/properties/questionType",
        "type": "Control",
      },
      Object {
        "label": "oie.security.question.createQuestion.label",
        "rule": Object {
          "condition": Object {
            "schema": Object {
              "const": "custom",
            },
            "scope": "#/properties/questionType",
          },
          "effect": "SHOW",
        },
        "scope": "#/properties/question",
        "type": "Control",
      },
      Object {
        "label": "oie.security.question.questionKey.label",
        "options": Object {
          "choices": Array [
            Object {
              "key": "eternal",
              "value": "What is love?",
            },
          ],
          "format": "dropdown",
        },
        "rule": Object {
          "condition": Object {
            "schema": Object {
              "const": "predefined",
            },
            "scope": "#/properties/questionType",
          },
          "effect": "SHOW",
        },
        "scope": "#/properties/questionKey",
        "type": "Control",
      },
      Object {
        "options": Object {
          "format": "password",
        },
        "scope": "#/properties/answer",
        "type": "Control",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
